File: code-playbook/sample_outbox_cdc_postgresql/src/main/java/com/revy/example/domain/outbox/OutboxEvent.java
package com.revy.example.domain.outbox;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "outboxevent")
public class OutboxEvent {

    @Id
    @Column(columnDefinition = "uuid")
    private UUID id;

    @Column(name = "aggregatetype", nullable = false)
    private String aggregateType;

    @Column(name = "aggregateid", nullable = false)
    private String aggregateId;

    @Column(nullable = false)
    private String type;

    @Column(nullable = false, columnDefinition = "text")
    private String payload;

    @Column(name = "created_at", nullable = false, insertable = false, updatable = false)
    private OffsetDateTime createdAt;

    protected OutboxEvent() {
    }

    private OutboxEvent(UUID id, String aggregateType, String aggregateId, String type, String payload) {
        this.id = id;
        this.aggregateType = aggregateType;
        this.aggregateId = aggregateId;
        this.type = type;
        this.payload = payload;
    }

    public static OutboxEvent orderCreated(Long orderId, String payload) {
        return new OutboxEvent(UUID.randomUUID(), "order", String.valueOf(orderId), "OrderCreated", payload);
    }

    public UUID getId() {
        return id;
    }

    public String getAggregateType() {
        return aggregateType;
    }

    public String getAggregateId() {
        return aggregateId;
    }

    public String getType() {
        return type;
    }

    public String getPayload() {
        return payload;
    }

    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }
}

File: pension/src/main/kotlin/com/example/pension/domain/OutboxEvent.kt
package com.example.pension.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.Table
import java.time.Instant
import java.util.*

@Entity
@Table(
    name = "outbox_event",
    indexes = [
        Index(
            name = "idx_outbox_unpublished",
            columnList = "published_at, created_at"
        )
    ]
)
class OutboxEvent protected constructor(
    @Column(name = "aggregate_type", nullable = false, length = 50)
    val aggregateType: String,

    @Column(name = "aggregate_id", nullable = false, length = 100)
    val aggregateId: String,

    @Column(name = "event_type", nullable = false, length = 100)
    val eventType: String,

    @Column(name = "payload", nullable = false, columnDefinition = "jsonb")
    val payload: String,

    @Column(name = "occurred_at", nullable = false)
    val occurredAt: Instant,

    @Column(name = "created_at", nullable = false)
    val createdAt: Instant,

    @Column(name = "published_at")
    var publishedAt: Instant?,

    @Column(name = "retry_count", nullable = false)
    var retryCount: Int,
) {
    @Id
    val id: UUID = UUID.randomUUID()

    fun markPublished(now: Instant) {
        publishedAt = now
    }

    fun markFailed() {
        retryCount += 1
    }

    companion object {
        fun create(
            aggregateType: String,
            aggregateId: String,
            eventType: String,
            payload: String,
        ) = OutboxEvent(
            aggregateType = aggregateType,
            aggregateId = aggregateId,
            eventType = eventType,
            payload = payload,
            occurredAt = Instant.now(),
            createdAt = Instant.now(),
            publishedAt = null,
            retryCount = 0,
        )
    }
}